vms/components/fee/gas.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"math"
+
+	"github.com/holiman/uint256"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+var maxUint64 = new(uint256.Int).SetUint64(math.MaxUint64)
+
+type (
+	Gas      uint64
+	GasPrice uint64
+)
+
+func (g Gas) AddPerSecond(gasPerSecond Gas, seconds uint64) Gas {
+	newGas, err := safemath.Mul64(uint64(gasPerSecond), seconds)
+	if err != nil {
+		return math.MaxUint64
+	}
+	totalGas, err := safemath.Add64(uint64(g), newGas)
+	if err != nil {
+		return math.MaxUint64
+	}
+	return Gas(totalGas)
+}
+
+func (g Gas) SubPerSecond(gasPerSecond Gas, seconds uint64) Gas {
+	gasToRemove, err := safemath.Mul64(uint64(gasPerSecond), seconds)
+	if err != nil {
+		return 0
+	}
+	totalGas, err := safemath.Sub(uint64(g), gasToRemove)
+	if err != nil {
+		return 0
+	}
+	return Gas(totalGas)
+}
+
+// MulExp returns an approximation of g*e^(excess / excessConversionConstant)
+//
+// This implements the EIP-4844 fake exponential formula:
+//
+//	def fake_exponential(factor: int, numerator: int, denominator: int) -> int:
+//		i = 1
+//		output = 0
+//		numerator_accum = factor * denominator
+//		while numerator_accum > 0:
+//			output += numerator_accum
+//			numerator_accum = (numerator_accum * numerator) // (denominator * i)
+//			i += 1
+//		return output // denominator
+//
+// This implementation is optimized with the knowledge that any value greater
+// than MaxUint64 gets returned as MaxUint64. This means that every intermediate
+// value is guaranteed to be at most MaxUint193. So, we can safely use
+// uint256.Int.
+//
+// This function does not perform any memory allocations.
+//
+//nolint:dupword // This comment is copied from the EIP-4844 specification
+func (g GasPrice) MulExp(
```

Comment:
Should this live in the `utils/math` pkg? It'll be reused for ACP-77's continuous fee mechanism

Reply:
I'd imagine ACP-77 will use this function directly